### ISSUE TITLE
Unexpected location in Parser caused by QP in predicate body

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#202]
Repro:
```Viper
field val:Int 

domain Array 
{
    function loc(a:Array, i:Int):Ref
}

predicate isArray(this:Array)
{
    forall i:Int :: acc(loc(this,i))
}
```

Stack Trace: 
```
11:14.471 ERROR: S: Error in backend: carbon stage: verify: java.lang.RuntimeException: unexpected location
11:14.471 S: at scala.sys.package$.error(package.scala:27)
	at viper.silver.parser.Translator.viper$silver$parser$Translator$$exp(Translator.scala:450)
	at viper.silver.parser.Translator.viper$silver$parser$Translator$$exp(Translator.scala:392)
	at viper.silver.parser.Translator$$anonfun$viper$silver$parser$Translator$$translate$8.apply(Translator.scala:95)
	at viper.silver.parser.Translator$$anonfun$viper$silver$parser$Translator$$translate$8.apply(Translator.scala:95)
	at scala.Option.map(Option.scala:146)
	at viper.silver.parser.Translator.viper$silver$parser$Translator$$translate(Translator.scala:95)
	at viper.silver.parser.Translator$$anonfun$7.apply(Translator.scala:44)
	at viper.silver.parser.Translator$$anonfun$7.apply(Translator.scala:44)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
11:14.471 S: at viper.silver.parser.Translator.translate(Translator.scala:44)
	at viper.silver.frontend.SilFrontend$class.doTypecheck(SilFrontend.scala:291)
	at viper.carbon.CarbonFrontend.doTypecheck(Carbon.scala:26)
	at viper.carbon.CarbonFrontend.doTypecheck(Carbon.scala:26)
	at viper.silver.frontend.DefaultFrontend$class.typecheck(Frontend.scala:193)
	at viper.carbon.CarbonFrontend.typecheck(Carbon.scala:26)
	at viper.server.ViperFrontend$class.execute(VerificationWorker.scala:107)
	at viper.server.ViperCarbonFrontend.execute(VerificationWorker.scala:423)
	at viper.server.VerificationWorker.startVerification(VerificationWorker.scala:56)
	at viper.server.VerificationWorker.run(VerificationWorker.scala:33)
	at java.lang.Thread.run(Thread.java:745)

```

[Issue created by arshavir: 2017-05-19]
[Last updated on bitbucket: 2018-08-13]

[Comment created by arshavir: 2017-05-19]
Of course ```acc(loc(this,i))``` is invalid syntax in Viper; therefore, this issue is about error reporting.

[Comment created by arshavir: 2017-05-19]
Detected by @a2scale.

Potentially related issue: https://bitbucket.org/viperproject/silicon/issues/287/predicates-with-qps-cause-stack-trace

[Comment created by mschwerhoff: 2017-11-26]
See also #177, #178, #214

[Comment created by vakaras: 2018-05-02]
Another instance:

```
#!silver

domain Array {
    function loc(): Ref
}
method test() {
    inhale acc(loc())
}
```

[Comment created by fabiopakk: 2018-08-13]
Issue fixed and committed to default branch.

**Comment by fpoli** [comment]
> **@vakaras** commented on 2017-05-26 14:40

One more example:

```
#!silver
            assume (forall s: Set[Ref], e1: Ref, e2: Ref ::
                {|s| == 1}
                (|s| == 1 && e1 in s && e2 in s) ==> e1 == e2)
```

**Comment by fpoli** [comment]
> **@vakaras** commented on 2017-05-28 17:34

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-06-19 14:41

As a related problem, triggers which don't mention all quantified variables are silently accepted. Carbon then generates invalid Boogie code, while Silicon appears to silently ignore the issue (I'm not sure how this works at the Z3 level). It seems that these could be consistency checks in the AST.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-06-19 14:43

Malte, could you check whether this is something you think fits Mousam's current work, please? I'm not sure if we have the utility methods for explicitly checking that a given trigger set is valid, but I could probably write them if necessary.

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-06-19 18:41

@alexander_summers looks like he did that; have a look at `class Trigger` and `class Forall` in pull request #73

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-08-25 14:48

Resolved by @mousam05 with pull request #73
